File: 0101-0200/0107-Binary Tree Level Order Traversal II/0107-Binary Tree Level Order Traversal II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> result;
        if (root == nullptr) {
            return result;
        }
        
        queue<TreeNode*> Q;
        Q.push(root);
        while (!Q.empty()) {
            int qs = Q.size();
            vector<int> path;
            for (int i = 0; i < qs; ++i) {
                TreeNode* node = Q.front();
                Q.pop();
                path.push_back(node->val);
                if (node->left) {
                    Q.push(node->left);
                }
                if (node->right) {
                    Q.push(node->right);
                }
            }
            result.push_back(path);
        }
        
        reverse(result.begin(), result.end());
        return result;
    }
};

File: 0101-0200/0107-Binary Tree Level Order Traversal II/0107-Binary Tree Level Order Traversal II.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        if root is None:
            return []
        result = []
        level = [root]
        while level:
            result.append([node.val for node in level])
            level = [child for node in level for child in (node.left, node.right) if child]
        return result[::-1]
